feat(tactic/clear): add clear' tactic

JLimperg opened the pull request:
Zulip: https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/A.20clear.20tactic.20that.20doesn't.20care.20about.20dependency.20order

We add an improved version of the `clear` tactic. When clearing multiple
hypotheses, `clear` can fail even though all hypotheses could be
cleared. This happens when the hypotheses depend on each other and are
given in the wrong order:

```lean
example : ∀ {α : Type} {β : α → Type} (a : α) (b : β a), unit :=
  begin
    intros α β a b,
    clear a b, -- fails
    exact ()
  end
```

When `clear` tries to clear `a`, `b`, which depends on `a`, is still in the
context, so the operation fails. We give a tactic `clear'` which
recognises this and clears `b` before `a`.

Please criticise away; I've tried to adhere to the various guidelines but have probably missed some stuff.

TO CONTRIBUTORS:

Make sure you have:

  * [x] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/contribute/naming.md)
  * [x] reviewed and applied [the documentation requirements](https://github.com/leanprover/mathlib/blob/master/docs/contribute/doc.md)
  * [x] for tactics:
     * [x] added or adapted documentation in [tactics.md](https://github.com/leanprover/mathlib/blob/master/docs/tactics.md)
     * [x] write an example of use of the new feature in [tactics.lean](https://github.com/leanprover/mathlib/blob/master/test/tactics.lean)
  * [x] make sure definitions and lemmas are put in the right files
  * [x] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

JLimperg committed:
feat(tactic/clear): add clear' tactic

We add an improved version of the `clear` tactic. When clearing multiple
hypotheses, `clear` can fail even though all hypotheses could be
cleared. This happens when the hypotheses depend on each other and are
given in the wrong order:

```lean
example : ∀ {α : Type} {β : α → Type} (a : α) (b : β a), unit :=
  begin
    intros α β a b,
    clear a b, -- fails
    exact ()
  end
```

When `clear` tries to clear `a`, `b`, which depends on `a`, is still in the
context, so the operation fails. We give a tactic `clear'` which
recognises this and clears `b` before `a`.

jcommelin commented on src/tactic/clear.lean:
```diff
@@ -0,0 +1,114 @@
+
+namespace native.rb_map
+
+  meta def index_map_aux {α} [hlt : has_lt α] [decidable_rel hlt.lt]
```
Minor style comment: contrary to old code, nowadays we don't indent code in `namespace`s. So this can go flush left.

JLimperg replied:
Thanks for the note. I've changed the indentation in the newest version.

cipher1024 reviewed (changes requested):
(see recommendation for conciseness and efficiency)

JLimperg committed:
refactor(tactic/clear): better implementation of clear'

We refactor `clear'`, replacing the old implementation with one that is
more concise and should be faster. The new implementation strategy also
gives us a new variant of `clear`, `clear_dependent`, almost for free.

`clear_dependent` works like `clear'`, but in addition to the given
hypotheses, it also clears any other hypotheses which depend on them.

JLimperg committed:
style(test/tactics, docs/tactics): less indentation

JLimperg committed:
test(test/tactics): better tests for clear' and clear_dependent

We make the tests for `clear'` and `clear_dependent` more meaningful:
They now permit less illegal behaviours.

JLimperg committed:
refactor(tactic/clear): simplify error message formatting

cipher1024 committed:
Merge branch 'master' into feat-smart-clear

mergify[bot] committed:
Merge branch 'master' into feat-smart-clear
